### ISSUE TITLE
Auth session sometimes ends with "Something went wrong trying to finish signing in."

### ISSUE THREAD

**Issue by PiotrSzlagura** [expo/expo#8957]
# Maintainer edit

**The auth session proxy service is deprecated and not recommended for production. For production apps, your app should navigate to the third-party authentication provider directly instead of using this service.** This is the official workaround to this issue, which is likely unresolvable due to how browser cookie policies have changed. Follow this [migration guide](https://github.com/expo/fyi/blob/main/auth-proxy-migration.md) if you are currently using the proxy service.

The migration guide shows how to configure the authentication provider to redirect directly to your app, typically with a deep link with your app's own URL scheme. In your app, set the useProxy option to false (the default) when calling the promptAsync method, which configures your app not to use this service.

Due to web browser changes like [WebKit's Tracking Prevention](https://webkit.org/tracking-prevention/), the AuthSession proxy service may not work reliably in edge cases such as when a user's device is configured to block cookies or prevent cross-site tracking. The AuthSession proxy service does not track nor collect any user data but it requires cookies to correctly redirect back to your app after the user has authenticated with the third-party authentication provider. This service will not work if the browser's settings or heuristics block cookies.

---

### Summary of Issue (just a few sentences)

Sometimes when I try logging in using AuthSession it doesn't redirect me back to app, but shows "Something went wrong trying to finish signing in." instead. After that when I try again, it works as it should. After numerous tests we assume that this only happens on iOS, it never got reproduced on Android.

I am using expo since version 33, and this never happened before.

### Environment - output of `expo diagnostics` & the platform(s) you're targeting

```
  Expo CLI 3.21.9 environment info:
    System:
      OS: macOS Mojave 10.14.4
      Shell: 5.3 - /bin/zsh
    Binaries:
      Node: 10.16.1 - /usr/local/bin/node
      Yarn: 1.15.2 - /usr/local/bin/yarn
      npm: 6.9.0 - /usr/local/bin/npm
    IDEs:
      Xcode: 10.2.1/10E1001 - /usr/bin/xcodebuild
    npmPackages:
      expo: ^37.0.0 => 37.0.11
      react: 16.9.0 => 16.9.0
      react-dom: 16.9.0 => 16.9.0
      react-native: https://github.com/expo/react-native/archive/sdk-37.0.1.tar.gz => 0.61.4
    npmGlobalPackages:
      expo-cli: 3.21.9
```

```
  Target env: Android & iOS
```

### Reproducible Demo

All I'm doing is:

```js
const {
  data: { redirect_url: redirectUrl }
} = yield ApiClient.get(`app.auth.${socialLoginSite}`); // fetch redirect url from backend, for google or facebook

const result = yield AuthSession.startAsync({
  authUrl: redirectUrl
}); // Everything happens in this call
```

### Steps to Reproduce

Try logging in using facebook or google, for some reason sometimes the final redirect doesn't work.

### Expected Behavior vs Actual Behavior

Expected Behavior: If there is an error, then AuthSession should return with

```
{
  result: {
    type: "error"
    ...
  }
}
```

Actual Behavior: AuthSession gets blocked on last step and both users and developers are confused.

**Comment by brentvatne** [comment]
hi there! we need some steps to reproduce this.

the source for the page you're seeing looks like this:

```jsx
export default function ExternalAuthSessionExitPage(props: Props): JSX.Element {
  const { debug, authProviderParams, returnURL } = props;
  return (
    <html>
      <head>
        <meta charSet="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        <title>Sign-in Complete</title>
      </head>
      <body style={ExternalAuthSessionStyles.body}>
        <main style={ExternalAuthSessionStyles.main as any}>
          {returnURL ? (
            <ReturnRedirect
              debug={debug}
              authProviderParams={authProviderParams}
              returnURL={returnURL}
            />
          ) : (
            <p>
              Something went wrong trying to finish signing in. Please close this screen to go back
              to the app.
            </p>
          )}
        </main>
      </body>
    </html>
  );
}
```

as you can see, if there is no returnURL available you will see this error message. the returnURL is extracted from cookies

**Comment by EvanBacon** [comment]
For future reference, I hit this error today when I misused the proxy. 

```tsx
const redirectUri = AuthSession.makeRedirectUri({
    useProxy: false,
});

// ...

promptAsync({ useProxy: false })
```

This caused the auth to open up to the provider (e.g. facebook.com) but redirect to the proxy `auth.expo.io` when it was complete. The error was the result of the auth proxy not being able to complete the result because I forgot to start it.

**Comment by dernster** [comment]
@EvanBacon I'm facing the same issue right now. I have an app in production and some users are facing this issue "sometimes". This is my code:

```
      const redirectUrl = AuthSession.getRedirectUrl();
      const result = await AuthSession.startAsync({
        returnUrl: redirectUrl,
        authUrl:
          `https://www.linkedin.com/oauth/v2/authorization` +
          `?response_type=code` +
          `&client_id=${LINKEDIN_CLIENT_ID}` +
          `&redirect_uri=${encodeURIComponent(redirectUrl)}` +
          '&scope=r_liteprofile%20r_emailaddress'
      });
```

Do you think I should change this code to use `promptAsync` instead?

**Comment by PiotrSzlagura** [comment]
@brentvatne I re-read documentation and updated my solution to look like this:

```js
const result = yield AuthSession.startAsync({
  authUrl: redirectUrl,
  returnUrl: `${Constants.linkingUrl}expo-auth-session`
});
```

In this auth session I always end up redirecting to `https://auth.expo.io/@company/slug?(...auth_data)` (ofcourse with right company and owner) and still sometimes it passes and sometimes "Something went wrong...". 

As you can see I pass `returnUrl` explicitly in this case, so condition in `ExternalAuthSessionExitPage` should not return false, but it seems it does. Any idea why? 

It's really hard to pinpoint specific situation this happens, as every test resolves in either working or not, with the same conditions (fresh install/relogging, multiple google accounts or single one, using facebook or google).

**Comment by PiotrSzlagura** [comment]
Ping @brentvatne Have you got any additional insight here?

**Comment by rrufus** [comment]
@brentvatne I also have this issue. I am using `expo-auth-session@~1.4.0` with Auth0 not LinkedIn as my provider, using Expo SDK 38.

I found that in Expo Client I never encountered an issue.
I found that when I start the managed app logged out from cold, the first time I attempt to log in, this issue occurs reliably. When I cancel the view and attempt to log in again, I can. Then when I log out and log in again, it works reliably.

It looks like in these specific circumstances there is some issue saving the returnUrl in the `auth.expo.io` layer. [Here](https://docs.expo.io/versions/v38.0.0/sdk/auth-session/#it-makes-redirect-url-allowlists-easier-to) it says:

`The Expo Auth service saves away the returnUrl (and if it is not a published URL or your registered custom theme, shows a warning page) and then sends the user off to the authUrl. When the authentication provider redirects back to https://auth.expo.io/@your-username/your-app-slug on success, the Expo Auth services redirects back to the returnUrl that was provided on initiating the authentication flow.`. 

I was using `Authsession.startAsync` with `authUrl`.
For `returnUrl` I was leaving it to go to the default, but if I specifically use `AuthSession.getDefaultReturnUrl()` nothing changes.
For the redirect_uri in `authUrl` I was using `AuthSession.getRedirectUrl()` but changing to `AuthSession.makeRedirectUri({ useProxy: true })` made no difference.

```
  Expo CLI 3.18.2 environment info:
    System:
      OS: macOS 10.15.7
      Shell: 5.7.1 - /bin/zsh
    Binaries:
      Node: 10.19.0 - ~/.nvm/versions/node/v10.19.0/bin/node
      Yarn: 1.22.4 - /usr/local/bin/yarn
      npm: 6.13.4 - ~/.nvm/versions/node/v10.19.0/bin/npm
    IDEs:
      Android Studio: 3.6 AI-192.7142.36.36.6392135
      Xcode: 12.1/12A7403 - /usr/bin/xcodebuild
    npmPackages:
      expo: ^38.0.0 => 38.0.9 
      react: 16.11.0 => 16.11.0 
      react-native: https://github.com/expo/react-native/archive/sdk-38.0.2.tar.gz => 0.62.2 
    npmGlobalPackages:
      expo-cli: 3.18.2
```

Can we re-open the issue?

My next step is to see if the issue still persists in SDK 39, since I am not sure how to debug this further from my side.

**Comment by brentvatne** [comment]
@rrufus - happy to reopen if you create a [mcve](https://stackoverflow.com/help/mcve) and share it. you can just create a new auth0 account specifically for this purpose so that we are sure we have the exact same setup

**Comment by rrufus** [comment]
@brentvatne ah this was my mistake, I used `expo init` and then picked the `minimal` workflow instead of the `blank` workflow.

However in the meantime, I have updated my iPhone X to 14.1, and I am not seeing the issue on the built TestFlight app, or on my production app.

So it seems the issue is isolated to 13.x, which is still an issue for our iOS users, but it's good to know there is a workaround.

**Comment by brentvatne** [comment]
interesting! that'll be helpful for reproducing. thanks @rrufus!

**Comment by jakeyr** [comment]
I am also experiencing this issue using Expo SDK 39, after deploying to app store. I'm using Auth0 and logs on the Auth0 show successful logins but failures on the expo redirect.

Additionally, I see the problem with users on iOS 14 (confirmed on the latest version, 14.2).

**Comment by jpagand** [comment]
Some of our users also experience this issue and cannot login back.
Our app is using  Expo SDK 39 but it happen from one day to another, not after an app release

So far it affect some users on iOS 14.1 and 14.2

**Comment by jakeyr** [comment]
Is there any way to view errors or logs for our application on auth.expo.io?

**Comment by jpagand** [comment]
Have someone found a workaround for users to login again? I currently have some users that see this message every time they try to login and that is quite critical

**Comment by rrufus** [comment]
@jpagand I have found that I am not receiving complaints of this issue now, and have not seen it myself for some time. However, nothing I have changed seems to obviously fix the issue. Regardless, here are the relevant things that were changed, in case one of them works for you:

1. In auth0, under "Branding", I switched experience from "Classic" to "New", and haven't seen issues since using "New".
1. In my logout process, I stopped doing a federated logout, meaning that auth0 rarely has to also handle authentication with another auth provider during login. **HOWEVER** the MCVE posted above never used a third party auth provider, only auth0's built in one.

**Comment by jakeyr** [comment]
@rrufus thanks for sharing. I am still currently experience the issue and I am using the "new" experience, and I do not do federated logouts.

**Comment by brentvatne** [comment]
hey folks. if you're having issues with this i'd recommend disabling the authsession proxy. we don't have time to investigate at the moment but you could build your own proxy server or provide multiple redirect uris

**Comment by jakeyr** [comment]
thanks @brentvatne. is the proxy open source by any chance?

**Comment by brentvatne** [comment]
it's not, but as a summary for how it works:

- you visit the start auth session page, we save the return url in cookies and send you to the auth page
- when you are redirected to the finish auth session page, get the return url from cookies, then extract the auth params from the url and append them to the return url and redirect

**Comment by jakeyr** [comment]
Update: I discovered that at least one user was having this issue due to having "Prevent Cross-Site Tracking" enabled under `Settings -> Safari` in the Privacy section. I am guessing this has to do with the way you are storing the redirect cookie being blocked somehow, or perhaps the JS on the redirect page not executing correctly while doing some cross-origin requests, etc.

Any chance you can take a look and see if there's a way to make this work now that you have something to go on?

**Comment by jpagand** [comment]
@jakeyr  trick to disable "Prevent Cross-Site Tracking" option also allowed our users to log in back

**Comment by jakeyr** [comment]
@brentvatne Here's what's confusing. In reverse engineering the proxy on my device, what I see is that production version of our app (on the app store) eliminates the return step through the proxy, and sends the deep link as the redirect URL directly to our OAuth provider in the `redirect_uri` parameter. So when I personally test the flow on our production app (not in the expo client), I don't hit the proxy on the way back to the app itself.

However, the issue my customers are seeing is that after completing sign-in with our provider, they land back on the proxy page and see the error. I assume the error has to do with the privacy setting on their device and the way cookies are being set (which is confusing, because it appears to be a normal cookie and there is a JS test that the cookie was set correctly on the `/start` interstitial page). So my questions are:

  1. Why do some customers seem to go through `auth.expo.io` on the way back to their apps, and some don't?
  2. For the customers who *do* go back through `auth.expo.io`, why does the `Prevent Cross-Site Tracking` setting in Safari prevent the proxy from working?

Thanks for any help!

**Comment by brentvatne** [comment]
> Why do some customers seem to go through auth.expo.io on the way back to their apps, and some don't?

this seems entirely up to the auth provider, they should be redirecting you back to the auth.expo.io url if configured correctly. we don't control where they redirect you to. you should be landing on the auth.expo.io page regardless of whether you're in expo client or in a standalone app unless you provided specific configuration for it not to do so, by disabling the proxy based on the context.

> For the customers who do go back through auth.expo.io, why does the Prevent Cross-Site Tracking setting in Safari prevent the proxy from working?

is this a reliable way to reproduce the problem? does this happen to you if you toggle the setting on your device?

**Comment by jakeyr** [comment]
> this seems entirely up to the auth provider

On second thought, I might have misread the redirect data when trying to replicate this on Safari on MacOS. Looks like I always go back through the proxy, which makes sense given the allowed return URLs I have configured on the auth provider.

> is this a reliable way to reproduce the problem? does this happen to you if you toggle the setting on your device?

unfortunately, no. I am unable to reproduce it even with the setting enabled. however, I can definitely find a user that can replicate the problem and can probably get you some client-side info if you need it.

**Comment by brentvatne** [comment]
> unfortunately, no. I am unable to reproduce it even with the setting enabled. however, I can definitely find a user that can replicate the problem and can probably get you some client-side info if you need it.

that would be helpful!

**Comment by jakeyr** [comment]
OK, for some reason this just started happening on my phone with the setting enabled. Here's what I'm seeing.

* The initial request to `/start` is met with a `304 Not Modified` response and a `Set-Cookie` response header with the `ru` cookie:

```
HTTP/1.1 304 Not Modified
Server: nginx/1.17.8
Date: Tue, 24 Nov 2020 20:26:14 GMT
Connection: keep-alive
Vary: Accept-Encoding, Origin
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
Set-Cookie: ru=muttville:///expo-auth-session; path=/@rowborg/muttville-bible; expires=Tue, 24 Nov 2020 21:26:14 GMT; secure
ETag: "afa-j+HpwY/8Yt2s5wGpgqfKdXU0y+E"
Strict-Transport-Security: max-age=31536000; includeSubDomains
```

* The auth completes and redirects to the proxy
* The request to the proxy complete page *does not* send the `ru` cookie in the request header--error ensues:

```
GET /@rowborg/muttville-bible HTTP/1.1
Host: auth.expo.io
Connection: keep-alive
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
User-Agent: Mozilla/5.0 (iPhone; CPU iPhone OS 14_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.1 Mobile/15E148 Safari/604.1
Accept-Language: en-us
Referer: https://muttville.auth0.com/u/login?state=g6Fo2SBGb2pQQTVjRVhYZExPZjc3QVllMnEzb3VjdzZXdFFmaKN0aWTZIGlhcUg1eTR4anJhZUl4TC1JNEczdVhIMDZPcVMxQWtSo2NpZNkgcXYyNk9xV3g5YkYwbXRUbmhkdDlvUU1NWWpERkJiMWk
Accept-Encoding: gzip, deflate, br
```

I suspect the issue is caused by a combination of the `Prevent Cross-Site Tracking` flag and the 304 cache response, leading to a cookie not being set?
```

**Comment by jakeyr** [comment]
@brentvatne Wondering if the meta `refresh` tag is redirecting before the cookie-checking JS is executing? And not sure why the 304 would somehow prevent the cookie from being set either.

**Comment by brentvatne** [comment]
@jakeyr - thanks for the info! do you happen to know what "Prevent Cross-Site Tracking" concretely does? i've tried googling it quickly and couldn't find a technical explanation

**Comment by jakeyr** [comment]
I'm not 100% sure what it does. It's a bit vague in the official description here:

https://webkit.org/tracking-prevention/

My best read is that it dynamically will adapt to being able to store cookies based on whether it thinks a webpage is being intentionally navigated to/from by the user. It makes reference to redirects:

```
Beyond across-the-board blocking of third-party cookies and downgrades of third-party referrers, ITP collects statistics on resource loads and matches it with known patterns of cross-site tracking. If a registrable domain matches at least one such pattern, it is classified as having cross-site tracking capabilities.

One such pattern is showing up as third-party resource under several first-party websites. A machine learning model decides when these three numbers leads to classification of domain.example:

The number of unique websites domain.example has been seen as third-party subresource under.
The number of unique websites domain.example has been seen as third-party iframe under.
The number of unique websites domain.example has been seen doing cross-site redirects under.
Another pattern that is detected as the capability to track cross-site is top frame redirects, often referred to as bounce tracking. ITP counts the number of unique such redirects that domain.example does, and classifies based on that number. ITP will count it as a bounce even if the redirect is delayed by landing on a webpage and triggering a navigation a couple of seconds later.
```

Wondering if the auto-redirect in a pop-up is somehow causing `auth.expo.io` to be classified as third-party?

**Comment by ide** [comment]
It might be just the fact that the frame redirects. We don't have much time to look into this but if you can figure out the root cause we can work on a fix.

AuthSession is just for convenience but there is nothing special about it. It's just code we wrote without using any special privileges in Expo. So you could write your own OAuth code and sidestep all of this if OAuth login is important to your app.

**Comment by tylerthedeveloper** [comment]
This may be related to the same issue but i am getting the following error only on simulator / in an IPA

```Error: No default redirect URL could be found. If you're using the bare workflow, you'll need to provide this yourself.```

As per the guides i have been using this and have had no problems until upgrading to big sur, upgrading npm, cleaning caches, etc.

```
  const [request, response, promptAsync] = AuthSession.useAuthRequest(
    {
      clientId: GITHUB_ID,
      clientSecret: GITHUB_SECRET,
      scopes: ['identity', 'user', 'public_repo', 'repo'],
      redirectUri: AuthSession.makeRedirectUri({
        // useProxy: true,
      })
      // redirectUri: 'https://auth.expo.io/@tcitrin/gitfolio'
    },
    discovery
  );
```

Expo diagnostics:

```
Expo CLI 4.0.15 environment info:
    System:
      OS: macOS 11.0.1
      Shell: 3.2.57 - /bin/bash
    Binaries:
      Node: 15.4.0 - ~/.nvm/versions/node/v15.4.0/bin/node
      npm: 7.0.15 - ~/.nvm/versions/node/v15.4.0/bin/npm
    Managers:
      CocoaPods: 1.9.1 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: iOS 14.2, DriverKit 20.0, macOS 11.0, tvOS 14.2, watchOS 7.1
    IDEs:
      Xcode: 12.2/12B45b - /usr/bin/xcodebuild
    npmPackages:
      expo: ^39.0.5 => 39.0.5 
      react: 16.13.1 => 16.13.1 
      react-dom: 16.13.1 => 16.13.1 
      react-native: https://github.com/expo/react-native/archive/sdk-39.0.3.tar.gz => 0.63.2 
      react-native-web: ~0.13.7 => 0.13.18 
      react-navigation: ^4.4.3 => 4.4.3 
    npmGlobalPackages:
      expo-cli: 4.0.15
    Expo Workflow: managed
```

**Comment by ide** [comment]
Over the years Apple has made some changes with how Safari handles cookies across websites. We've currently got a long queue of other things we're focused on but as with most issues, if someone would like to investigate this more and identify concretely what the root cause is and how to fix it, that would be helpful.

**Comment by ide** [comment]
If this is important enough to anyone, we could use help figuring out what the root cause is -- this issue is marked as "Help Wanted." It will probably require hooking up the JS inspector to mobile Safari, writing test sites / proofs of concept, and identifying what Safari's behavior exactly is.

**Comment by ivansenic** [comment]
We are seeing this issue in the production on iOS with Safari's `Prevent Cross-Site Tracking`. What's really anoying is that this has been happening a while and some amount of users where really pissed, as they can not login in the app. Now we discovered this is related exactly to this issue.

@ide I think this is way more critical than "if this is important enough to anyone".. You basically have no idea what browsers will be used to open the login screen and what the settings are. This means that the login can consistently fail for good amount of people. How is this not a big issue?

It's also a problem that there is no actual workaround for this. If you continue using the `AuthSession` with `proxy: true` you will hit this problem and there's no way out. Or am I missing something?

**Comment by ide** [comment]
The most helpful thing is to figure out the root cause and how to sustainably address it.

**Comment by ivansenic** [comment]
@ide So can you just tell me, since you have been longer in the discussion, if this problem occurs also if `useProxy: false` is used.. I would guess not as there is no proxy in the middle then? I f this is the case then at least there is a workaround to not use the proxy in the standalone env. Am I right?

**Comment by ide** [comment]
@ivansenic I don't know the answer to that question. If setting `useProxy: false` were to have an effect, it would be if Safari's cross-site tracking prevention kicks in when the browser visits the intermediate redirect page but not when visiting the OAuth provider directly. That seems like a reasonable hypothesis.

**Comment by erikkjernlie** [comment]
We are having the exact same issue - some users cannot log in. So you have any updates here?

**Comment by Dakuan** [comment]
same same

**Comment by ivansenic** [comment]
I can confirm that we managed to overcome this by not using the proxy in the production or better said when not using the expo client. We confirmed that the no-proxy solution works with the clients where we were able to reproduce the problem. We kept the proxy for expo client as it's easier to handle dev and integration stages like this, but you can also fully move to no-proxy solution. 

In addition, we had to move away from `AuthSession.startAsync` to `AuthSession.loadAsync` and `AuthRequest.promptAsync`. The reason is that we needed to supply correct information about proxy usage to both `makeRedirectUri` and `promptAsync`.. 

Basically this worked for us:

```
export const isAuthSessionUseProxy = () => Constants.appOwnership === AppOwnership.Expo;

export const getAuthSessionRedirectUrl = () => AuthSession.makeRedirectUri({ native: 'com.myapp://my-redirect-path', useProxy: isAuthSessionUseProxy() });

export const fetchToken = () => {
  // maybe complete auth session first
  WebBrowser.maybeCompleteAuthSession();
  
  // resolve redirect url
  const redirectUrl = getAuthSessionRedirectUrl();
  
  const discovery = {
      authorizationEndpoint: `${Env.AUTH_HOST}/oauth/authorize`,
      tokenEndpoint: `${Env.AUTH_HOST}/oauth/token`
  };
  
  return AuthSession.loadAsync(
      {
          clientId:..,
          redirectUri: redirectUrl,
          extraParams: {
              // anything else you want to pass to authorize endpoint as request param 
          }
      },
      discovery
  )
  .then(request => request.promptAsync(discovery, { useProxy: isAuthSessionUseProxy() }))
  .then(result => { 
     // get the token with the code if the result is successful 
     // don't forget to use the same redirectUrl
   });
}
```

**Comment by JohnEmerson1406** [comment]
https://docs.expo.io/guides/authentication/#github

I did exactly as it is in the documentation, the browser opens, I can log in successfully, but instead of being redirected to the app, I get “Something went wrong trying to finish signing in. Please close this screen to go back to the app.”

**Comment by brentvatne** [comment]
if you have encountered this problem, please do not use the proxy option in authsession. it is not necessary, it's just a helper and if it's not helping then don't use it :)

we will investigate in a future sdk, we don't have time for it for sdk 41. if someone would like to investigate further and summarize your findings or open a pr, that would be welcome!

**Comment by phatmann** [comment]
The proxy server still has this issue on iOS

**Comment by physixfan** [comment]
Our app also have this issue. 

I see the error frequently in my Xfinity home Wi-Fi internet environment, but this error does not appear in my AT&T mobile phone internet environment.

I see the error when I try to log in a new user, but when i try the same user the second time, the error generally does not appear.

We have tried to turn off the proxy option, as well as turn on the “allow cross-site tracking” on the users’s side. Neither of them is useful.

**Comment by tlmader** [comment]
Using the proxy server is a recommended workaround to another issue with AuthSession returning `dismiss` result on Android Standalone: https://github.com/expo/expo/issues/12044.

However, that workaround lands me on this issue.

**Comment by tlmader** [comment]
> if you have encountered this problem, please do not use the proxy option in authsession. it is not necessary, it's just a helper and if it's not helping then don't use it :)
> 
> we will investigate in a future sdk, we don't have time for it for sdk 41. if someone would like to investigate further and summarize your findings or open a pr, that would be welcome!

Using the proxy server is a recommended workaround to another issue with AuthSession returning dismiss result on Android Standalone: #12044. However, that workaround lands me on this issue. I currently have no path forward for implementing sign in for Android standalone. Any plans to look at this for SDK 43?

**Comment by ide** [comment]
We currently don't have a known fix for this due to how Apple has changed Safari's behavior over the years in response to internet ads and changes in the data tracking landscape. Google may have changed Android/Chrome as well. The little AuthSession server that Expo runs performs no tracking and has no ads obviously but may have been a casualty in Apple & Google's changes to the web.

**Comment by Jeysoon** [comment]
I was having exactly the same issue just with one user who reported it, and some others happening randomly or if they tried the second time it will disapear, all this was fixed as @jakeyr said, you just go Settings -> Safari under the Privacy section and toggle off "Prevent Cross-Site Tracking", that worked for this user i mention who had rougly a year without being able to log in because of this, just as a side note im using SDK 42, hope this helps someone out there.

**Comment by jessejonass** [comment]
Well, in my case, the problem was solved by signing directly into Expo CLI.
https://docs.expo.dev/workflow/expo-cli/

As a guarantee, I deleted the state.json file which in my case is located in C:\Users\MyUser\.expo

**Comment by davyengone** [comment]
Still having the same issue here. Posting here to keep the thread open! 

`Expo CLI 5.0.3 environment info:
    System:
      OS: macOS 12.2
      Shell: 5.8 - /bin/zsh
    Binaries:
      Node: 16.13.1 - /usr/local/bin/node
      Yarn: 1.22.15 - ~/.yarn/bin/yarn
      npm: 8.1.2 - /usr/local/bin/npm
    Managers:
      CocoaPods: 1.10.1 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 21.2, iOS 15.2, macOS 12.1, tvOS 15.2, watchOS 8.3
    IDEs:
      Android Studio: 2020.3 AI-203.7717.56.2031.7935034
      Xcode: 13.2.1/13C100 - /usr/bin/xcodebuild
    npmPackages:
      expo: ^44.0.2 => 44.0.2 
      react: 17.0.1 => 17.0.1 
      react-dom: 17.0.1 => 17.0.1 
      react-native: 0.64.3 => 0.64.3 
      react-native-web: 0.17.1 => 0.17.1 
      react-navigation: ^4.4.4 => 4.4.4 
    npmGlobalPackages:
      expo-cli: 5.0.3
    Expo Workflow: managed
`

**Comment by busyxiang** [comment]
I am facing this issue recently, my ios build cannot pass Apple Review at all because of this issue, and most frustrating is I tested the testflight build only my physical iPhone device and was never able to trigger the issue myself. I am super confused right now as to how to resolve this issue

**Comment by busyxiang** [comment]
> Hi, all
> 
> We've been suffering from this problem for a while now. We are using auth0 as our oauth service provider.   We found that the first login always fails if the user's network connection is under certain conditions. The guess is that it is a network connection problem between the expo auth proxy server and the auth0 auth server. At last, our solution is: skipping the expo auth server and callback to the our app directly. Now, the error never happens again and everything works fine.
> 
> Here is the code:
> 
> ```
> function login() {
>   const redirectUrl = 'app_schema://callback/url/';
>   let urlParams = {
>     client_id: auth0ClientId,
>     response_type: 'token',
>     scope: 'openid profile email',
>     redirect_uri: redirectUrl,
>   };
>   let loginParams = {
>     authUrl: `${auth0Domain}/authorize` + toQueryString(urlParams),
>     returnUrl: redirectUrl
>   }
>   return AuthSession.startAsync(loginParams);
> }
> ```

This is the solution that works for me

**Comment by cheejiayuan512** [comment]
Hi guys, if anyone is facing this issue on android, do check that you do not have any adblockers/AdGuard enabled. It recently started blocking the returnURI and prevents the final redirect back to the app.

**Comment by conoremclaughlin** [comment]
We're also facing this in current development while upgrading to v45. Not really sure why the working [expo-google-app-auth](https://github.com/expo/expo/tree/sdk-44/packages/expo-google-app-auth) was deprecated in favor of a library that doesn't work all the time for iOS and Android?

https://github.com/expo/expo/blob/sdk-44/docs/pages/versions/v44.0.0/sdk/google.md

We're attempting some of the fixes above but longer support of expo-google-app-auth would also help.

**Comment by blin17** [comment]
having this issue too where i cannot login-- very frustrating

**Comment by hyochan** [comment]
Facing this when trying to login with `github`. It works without a proxy.

**Comment by whyaftab** [comment]
This package in unstable I waisted whole day and at last downgraded expo to 44 and installed expo-google-app-auth which only took 20 mins to fix this issue.

**Comment by tsaiDavid** [comment]
> > Hi, all
> > We've been suffering from this problem for a while now. We are using auth0 as our oauth service provider.   We found that the first login always fails if the user's network connection is under certain conditions. The guess is that it is a network connection problem between the expo auth proxy server and the auth0 auth server. At last, our solution is: skipping the expo auth server and callback to the our app directly. Now, the error never happens again and everything works fine.
> > Here is the code:
> > ```
> > function login() {
> >   const redirectUrl = 'app_schema://callback/url/';
> >   let urlParams = {
> >     client_id: auth0ClientId,
> >     response_type: 'token',
> >     scope: 'openid profile email',
> >     redirect_uri: redirectUrl,
> >   };
> >   let loginParams = {
> >     authUrl: `${auth0Domain}/authorize` + toQueryString(urlParams),
> >     returnUrl: redirectUrl
> >   }
> >   return AuthSession.startAsync(loginParams);
> > }
> > ```
> 
> This is the solution that works for me

Curious how you're handling logout? We're only experiencing this when trying to logout and using `openAuthSessionAsync` from `expo-web-browser`.

**Comment by Oao2021** [comment]
Try following the documentation here [( https://docs.expo.dev/guides/authentication/#google)]

It may help to follow the screen shot there.

Not sure if it was what helped solve the bug in our case, where in some devices the google sign in was not completing. We added the Web dev **https://localhost:19006/** to our hosts URLs and also the **https://localhost:19006/**  to the Redirect URLs. So far so good.

**Comment by Oao2021** [comment]
Try following the documentation here https://docs.expo.dev/guides/authentication/#google

It may help to follow the screen shot there.

Not sure if it was what helped solve the bug in our case, where in some devices the google sign in was not completing. We added the Web dev https://localhost:19006/ to our hosts URLs and also the https://localhost:19006/ to the Redirect URLs. So far so good.

**Comment by Oao2021** [comment]
Try following the documentation here https://docs.expo.dev/guides/authentication/#google

It may help to follow the screen shot there (under the Firebase option) if you are using Firebase

Not sure if it was what helped solve the bug in our case, where in some devices the google sign in was not completing. We added the Web dev https://localhost:19006/ to our hosts URLs and also the https://localhost:19006/ to the Redirect URLs. So far so good.

**Comment by charlestbell** [comment]
Wow. A year later and it's unfixed. Wasted 3 days of my time. It works fine on ios, but on android AuthSession is still broken.

 I'll try switching to expo-google-app-auth and see how that goes.

**Comment by LorraneRochaS** [comment]
I have the same problem, has anyone found a solution?

**Comment by ide** [comment]
The official solution is to configure the third-party auth providers to directly link to your app instead of using the intermediate proxy service. This is more secure (see this [security advisory](https://blog.expo.dev/security-advisory-for-developers-using-authsessions-useproxy-options-and-auth-expo-io-e470fe9346df)) and more reliable (does not have cookie issues discussed in this thread). 

**How to migrate:** a migration guide provided [here](https://github.com/expo/fyi/blob/main/auth-proxy-migration.md) explains how to switch from using the `useProxy` option, deprecated as of SDK 48, to directly linking to your app's own callback URL. If you are using Expo Go to develop, you will also need to graduate to using a [development build](https://docs.expo.dev/development/introduction/) of your own app, which lets you specify custom URL schemes and universal link domains in your app.

You can still use the [`AuthSession`](https://docs.expo.dev/versions/latest/sdk/auth-session/) API without the `useProxy` option. Be sure to specify your app's own callback URL directly, which could be a deep link with your custom URL scheme or a universal link if you have a domain associated with your app.
